x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RequestXContent.java

```diff
@@ -216,11 +243,146 @@ private static QueryParams parseParams(XContentParser p) throws IOException {
+        String paramKeyName,
+        List<XContentParseException> errors,
+        XContentLocation loc
+    ) {
+        ParamParsingKeys paramType = null;
+        try {
+            paramType = ParamParsingKeys.valueOf(paramKeyName.toUpperCase(Locale.ROOT));
+        } catch (IllegalArgumentException iae) {
+            errors.add(
+                new XContentParseException(
+                    loc,
+                    "["
+                        + paramKeyName
+                        + "] is not a valid param attribute, "
+                        + "a valid attribute is any of "
+                        + Arrays.stream(ParamParsingKeys.values())
+                            .map(p -> p.toString().toLowerCase(Locale.ROOT))
+                            .collect(Collectors.joining(", "))
+                )
+            );
+        }
+        return paramType;
+    }
+
+    private static ParserUtils.ParamClassification getClassificationForParam(
+        String paramName,
+        HashMap<ParamParsingKeys, Object> paramElements,
+        XContentLocation loc,
+        List<XContentParseException> errors
+    ) {
+        ParserUtils.ParamClassification classification = ParserUtils.ParamClassification.CONSTANT;
+        Object value = paramElements.get(ParamParsingKeys.VALUE);
+        // paramAttributes(0) = true if this is set as an identifier, paramAttributes(1) = true if this is set as an identifier pattern
+        BitSet identifierOrPattern = new BitSet(2);
+        var id = paramElements.get(ParamParsingKeys.IDENTIFIER);
+        if (isValidateIdentifierOrPatternField(id, ParserUtils.ParamClassification.IDENTIFIER, loc, errors)) {
+            identifierOrPattern.set(0);
+            classification = ParserUtils.ParamClassification.IDENTIFIER;
+        }
+        var pattern = paramElements.get(ParamParsingKeys.IDENTIFIERPATTERN);
+        if (isValidateIdentifierOrPatternField(pattern, ParserUtils.ParamClassification.PATTERN, loc, errors)) {
+            identifierOrPattern.set(1);
+            classification = ParserUtils.ParamClassification.PATTERN;
+        }
+        // a param cannot be both identifier and pattern
+        if (identifierOrPattern.cardinality() > 1) {
+            errors.add(
+                new XContentParseException(
+                    loc,
+                    format(
+                        null,
+                        "[{} = {value={}, {}=true, {}=true}] can be either an {} or an {}, but cannot be both",
+                        paramName,
+                        value,
+                        ParamParsingKeys.IDENTIFIER.toString().toLowerCase(Locale.ROOT),
+                        ParamParsingKeys.IDENTIFIERPATTERN.toString().toLowerCase(Locale.ROOT),
+                        ParamParsingKeys.IDENTIFIER.toString().toLowerCase(Locale.ROOT),
+                        ParamParsingKeys.IDENTIFIERPATTERN.toString().toLowerCase(Locale.ROOT)
```

Comment:
since this occurs in multiple places, how about creating a lowerCase name method on the enum and calling that instead.